frontends/p4/frontend.cpp

```diff
@@ -172,11 +173,17 @@ const IR::P4Program *FrontEnd::run(const CompilerOptions &options, const IR::P4P
         new ResolveReferences(&refMap),  // check shadowing
         new Deprecated(&refMap),
         new CheckNamedArgs(),
+        new DeclarationCopyPropagation(&refMap, &typeMap, true),
```

Comment:
Can you please explain it a bit further?
I can create a class that contains the constants and will be the parent of the TypeMap class, but I would still need to use the typeMap in DeclarationCopyPropagation so that TypeInference could see them.

Reply:
The information will be used in that TypeInference when checking if variables are compile time constant. This is needed only here because the DeclarationCopyPropagation on line 176 won't propagate any constants. The propagation is done in the PassRepeated DeclarationCopyPropagation, after the first TypeInference has passed.

Reply:
I see. That could be a problem is additional passes are inserted in between that information could be lost.
I wonder whether a separate data structure should be used for that purpose (but that's not easy).
A second problem is that any changes in the IR between these two passes may invalidate this information: if a child of any of the nodes you analyze will change, the nodes will be replaced by clones, so they won't be found in the TypeMap.
Ideally you should propagate the constants right away, immediately after you detect them. But I assume you need type information. Can't you put the analysis and propagation next to each other?

Reply:
I intended that the passes in between, including DeclarationCopyPropagation, would be viewed as a group and not changed, but you are right, additional passes could be inserted.
Yes, first it was the plan to do everything right away, but I needed TypeInference to check types and do other work, and only then propagate. The problem was with compile time constant errors, so I needed to split the pass calls to solve this.

Now that I think of it, maybe a good solution would be to add a skipCompileTimeConstantsCheck flag in TypeInference. That way I could postpone the compile time constant check, and DeclarationCopyPropagation on line 176 wouldn't be needed anymore, everything could be done in PassRepeated afterwards. Also, in that case, the use of TypeMap could be completely avoided.

Reply:
Type checking uses the constant property to check that some expressions are compile-time constants, e.g., constructor parameters. It may be possible to split this functionality in a separate pass, but it may not be easy.

Your proposal to avoid giving errors in the first type-checking pass is a possibility. The type-checking pass is accumulating a few such boolean flags, which will make the code harder to understand, but they are relatively local.
If you make this change we should move from having boolean arguments in the constructor, that is completely unreadable, please define some enums with the options for typechecking.

```
enum class TypeCheckingOptions {
   SkipConstantCheck,
   ReadOnly,
}
```

We could put these in the typeMap - in fact, this is what setStrictStruct does.

Reply:
Thank you for your advice! I will try to do that and reply when I'm done.